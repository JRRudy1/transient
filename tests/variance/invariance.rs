//! Ensures that a function signature requiring covariant use of an invariant
//! wrapper gets rejected

#![allow(dead_code)]
use transient_any::*;


// This function requires `long` to shorten its lifetime from 'b to 'a, and
// should be *rejected* if the type is considered to be *invariant*.
fn shrink<'a, 'b: 'a>(long: ErasedInv<'b>, _short: &'a String) -> ErasedInv<'a> {
    long
}

fn main() {
    // This test should fail to compile
}