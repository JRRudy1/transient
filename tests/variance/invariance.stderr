error: lifetime may not live long enough
  --> tests/variance/invariance.rs:11:5
   |
10 | fn shrink<'a, 'b: 'a>(long: ErasedInv<'b>, _short: &'a String) -> ErasedInv<'a> {
   |           --  -- lifetime `'b` defined here
   |           |
   |           lifetime `'a` defined here
11 |     long
   |     ^^^^ function was supposed to return data with lifetime `'b` but it is returning data with lifetime `'a`
   |
   = help: consider adding the following bound: `'a: 'b`
